misc/jobserver_pool.py

```diff
@@ -0,0 +1,340 @@
+
+# Start <command> with a fixed number of jobs
+/path/to/jobserver_pool.py -j10 <command>
+
+# Disable the feature with a non-positive count. This is equivalent
+# to running <command> directly.
+/path/to/jobserver_pool.py -j0 <command>
+
+# Use a specific FIFO path
+/path/to/jobserver_pool.py --fifo=/tmp/my_build_jobs <command>
+
+# Setup jobserver then start new interactive Bash shell
+# session, print MAKEFLAGS value, build stuff, then exit.
+/path/to/jobserver_pool.py bash -i
+echo "$MAKEFLAGS"
+... build stuff ...
+exit
+"""
+        )
+        return 0
+
+    def check_pipe_tokens(read_fd: int, jobs_count: int) -> int:
+        if jobs_count <= 1:  # Nothing to check
+            return 0
+
+        # Remove implicit token from the expected count.
+        expected_count = jobs_count - 1
+        os.set_blocking(read_fd, False)
+        read_count = 0
+        while True:
+            try:
+                token = os.read(read_fd, 1)
+                if len(token) == 0:  # End of pipe?
+                    break
+                read_count += 1
+            except BlockingIOError:
+                break
+
+        if read_count < expected_count:
+            print(
+                f"ERROR: {expected_count - read_count} tokens were missing from the jobs pool (got {read_count}, expected {expected_count})",
+                file=sys.stderr,
+            )
+            return 1
+        if read_count > expected_count:
+            print(
+                f"ERROR: {read_count - expected_count} extra tokens were released to the jobs pool (got {read_count}, expected {expected_count})",
+                file=sys.stderr,
+            )
+            return 1
+
+        return 0
+
+
+def main() -> int:
+    parser = argparse.ArgumentParser(
+        description=__doc__, formatter_class=argparse.RawTextHelpFormatter
+    )
+    if _IS_WINDOWS:
+        parser.add_argument(
+            "--name",
+            help=f"Specify semaphore name, default is {_DEFAULT_NAME}",
+            default=_DEFAULT_NAME,
+        )
+    else:
+        mutex_group = parser.add_mutually_exclusive_group()
+        mutex_group.add_argument(
+            "--pipe",
+            action="store_true",
+            help="Implement the pool with a Unix pipe (default is FIFO).",
+        )
+        mutex_group.add_argument(
+            "--fifo",
+            default=_DEFAULT_NAME,
+            help=f"Specify pool FIFO file path (default ./{_DEFAULT_NAME})",
+        )
+
+    parser.add_argument(
+        "--no-check",
+        action="store_true",
+        help="Disable the final check that verifies that all job slots were returned to the pool on exit.",
+    )
+
+    parser.add_argument(
+        "--help-usage", action="store_true", help="Print usage examples."
+    )
+
+    parser.add_argument(
+        "-j",
+        "--jobs",
+        action="store",
+        metavar="COUNT",
+        dest="jobs_count",
+        type=int,
+        default=os.cpu_count(),
```

Comment:
That was the wrong  `cpu_count` - it does not correctly reflect which CPU cores are masked for this process. `os.process_cpu_count` (for Python 3.13 and above)  gives the correct available core count.